### DISCUSSION TITLE
Future of Java UI

### DISCUSSION THREAD

**Discussion by bevans2000** [mars-sim/mars-sim#979] [General]
There have been a few mentions of what to do with the UI long term such as create a Web UI, move to JavaFX. These relate to the technology used but the first question is how should a user interact with the UI. @mokun & @Urwumpe 

### User Interaction
The current user experience with the simulation can be confusing as the user is presented with a desktop and a number of tools. What is the best way to interact with the simulation? It is not obvious although the UI is very information rich which is good. The current approach allows the user to interact in any way.

An alternative is to adapt a more structured experience, consider for example Eclipse or VSCode where the user has a layout pre-defined. This workbench style would be very suitable from mars-sim. There is a mock-up below.

![37894662dd6747f587f0edb4bc22e621](https://github.com/mars-sim/mars-sim/assets/20720166/92f27247-5e36-436e-905d-9dd202cb681b)
The components are:

- Left hand panel is a tree list displaying all entities grouped according to the type, e.g. Person, Robots, Mission etc, parented to a Settlement. 
- Left bottom panel shows the brief details of the selected entity in the tree. Only a few basic attributes.
- Bottom panel show tables of entities, this is similar to the existing MontiorTool. As a folder node is selected in the tree; the table is populated with the child entities. For example the Persons in a Settlement. Tabs can be pinned so they remain fixed otherwise the tab is reused as new selections are done.
- Bottom Panel Console - The existing console interface would now be a tab within the UI instead of a separate window outside.
- Right panel shows constant tools
- Right panel Time tab - The existing Time tool shown in a tab
- Right Panel Event tab - A feed (list) of Events that are shown. This has filter buttons at the top. This feed has a maximum number of events displayed.
- Center panel is a desktop where Unit Windows are rendered. These are generated whenever a Unit is select in the Event, Monitor or Tree panel.

Missions, Science & Supply Missions are now rendered as standalone windows in this desktop. The Mission, Science and Transport tools are no longer used because the entities can be selected in the tree/table/event feed. The right panel viewing panels of these tools become the windows shown in the center panel.

The tree panel can also support a Favourites as well.

### Implementation
The implementation is not as easy. With the above adopting the Eclipse workbench framework would be a good idea BUT this means adopt SWT which would need a total rebuild of all UI panels. So this would involve a lot of effort.
The same applies for JavaFX in that nothing can be re-used. 

A web UI I feel is the ultimate goal and whilst this can adopt the workbench style it involved even more development effort and bigger question on how a web API should be developed.

Therefore the Swing implementation is likely to be our best opinion.

**Comment by Urwumpe** [comment]
I agree on swing, we used it so far and I see no explicit reason, why swing is no longer enough. What we could consider is using Radiance for Look-And-Feel and some nicer ways to handle actions. But that is rather maintenance, it alone does not provide new features or better usability.

I think, the real question for the UI is, which goals we want to pursue with the UI in the future. Personally, I am not a fan of the classic Multiwindow approach, when it comes to usability. IMHO, better would be a good navigation breadcrump with a single main focus window and one or two detail windows at the sides or at one side, providing additional information. A bookmark UI component (tab? menu?) could be nice to save often used views (with their focus object, e.g. "Processes: Alpha Base") The views should be configureable to the needs of the player.  

Also possible: Saving an invocation context to a view configuration. Like, in prosa: User is clicking on a button next to a settlement in the event view for a medical emergency and the views are opened to a medical context with the settler mentioned by the event. The player could choose to open the settlements medical overview in the main view, while the medical data of the settler alone is displayer in a side view. It might be possible to select that a side view should show the view from which this view configuration is invoked, e.g. the event view.

**Reply by bevans2000** [reply]
I agree the current multi-window is not ideal but the simulation has concurrent actions occurring so a single window approach may be limiting?
Nevertheless we need to consider changing something.

**Reply by Urwumpe** [reply]
Well, most humans and players are bad at multitasking. You just look at one window for a while and then switch it, if you switch too fast, you can't do much. For additional data, there is the idea with the secondary view. 

Even if we think about aircraft, the pilots rarely have more than two major displays each. Not sure, if it would make sense for larger or wider screens, to permit two main views with their associated context and secondary screens to be displayed.

Like that, (sorry, I made this quickly and not for accuracy or beauty)
![WideScreenViewDoodle](https://github.com/mars-sim/mars-sim/assets/10480792/019183c9-9c58-4778-99ab-2f65f5f837a5)

**Reply by mokun** [reply]
> most humans and players are bad at multitasking. 

Good point ! 

It doesn't feel like having too much freedom in opening whatever Unit Windows in the desktop can become counterproductive. 

They've done studies on how living space is related to the brain developed of mice. The first group of mice are initially given a huge area to roam around freely without restriction. The second group of mice are initially given only a smaller area to live but gradually having more space opened up for forage. The second group turns out to be smarter and having more developed front lobe to respond to the environment.

I suppose this could be applied to human being :D. If not, I do see it's a better approach for mars-sim in having more structural UI with just one main view and perhaps a second view and some side views if needed.

**Reply by bevans2000** [reply]
I think we are all actually agreeing. In the center "main" view there would only ever be one window visible. So there would only be 4 oanels visible but in the center the user has tabs to flip between different Entity Windows.

**Reply by Urwumpe** [reply]
Sounds OK for me, but I would include general or "planet-wide" views as well. For keeping track what happens, we could simply use a small log view at the side, maybe with configurable triggers, about what kind of events the player wants to be notified there. Nice would be some non-text indicators, of what severity the events are, so that its easy to see if a emergency happens.

Again, for a wide screen setup or for people who use two displays or more, we could allow an alternative UI. For example, we could provide a Top-level Frame, which just shows a view. 

Its some time to christmas, but I ask: Can I have a rotating and shaded view of Mars in JOGL like a screen-saver, which just switches between certain information boxes or highlights the position of a settlement or rover, while time passes? Or maybe shows the latest event when its location is visible for the player?  With Earth time and the current Mars time on top? Maybe we could also see the trajectories of incoming or departing spacecraft there. 

Would look nice in full screen and could catch some attention.

**Reply by mokun** [reply]
>  for a wide screen setup or for people who use two displays or more, we could allow an alternative UI. For example, we could provide a Top-level Frame, which just shows a view.

That's true. I always run mars-sim on a machine with 2 LCDs set up.

For now, I usually use one LCD in Eclipse for debugging and the other LCD for running mars-sim's Swing UI. 

And I put mars-sim's console on either LCD.

btw, @bevans2000, can you modify UIConfig to save the position of the mars-sim console/terminal ?

> Would look nice in full screen and could catch some attention.

One problem is that the concept of full screen in swing is ill-defined and is not the same as the full screen mode in some games. In JavaFX, there's a full screen mode but it may or may not be what we want...

**Comment by mokun** [comment]
@bevans2000, 

Is it possible we use the mars-sim.com site for bouncing off JSON messages ? 

Say, I host mars-sim server version at home. It will listen in any JSON messages via the mars-sim.com site.

In the mars-sim server, I may be running 2 different simulation sessions. The first session is an old build that's more stable. The other one is a newer, unstable build for testing features.

Then you run mars-sim client version in your house and you log onto the new build to test it.

At the same time, @Urwumpe runs mars-sim client in his house and log onto the old build to play mars-sim.

Any networking library out there we may look into to achieve this ?

**Reply by bevans2000** [reply]
Yes this is a classic architecture and not difficult. The problem is our current ui is tightly coupled to the core engine, there is no interface between the 2. So how would a  client make a remote call to get the details of a Person for example?
I've experimented with a web client. As we have an object graph in our engine GraphQL is the right choice for our remote API. Doing that part is hard but not impossible. The ui I see as a problem.

**Reply by mokun** [reply]
May I ask why and in what ways you see GraphQL being a good choice ? Will it be too complicated ?

**Reply by bevans2000** [reply]
GraphQL is very good at exchanging partial object models which is what our simulation engine is based on. The UI screens use a hierarchy of objects to render a screen screen. Traditional REST approach involves a lot of calls per screen. 
There is a lot of You Tube explaining the benefits

GraphQL has a lot of libraries out of the ox now, particularly in SpringBoot.

**Comment by bevans2000** [comment]
Thinking about #1149 I've realised that tools like the Science tool could be redesigned to be consistent with the Unit window approach. So Science studies become a tab in the Monitor Tool and they have a dedicated Window when selected.
This is moving towards the pattern of:

- Single central Monitor tool showing overview
- Detailed view in a dedicated window

The latter would be the central panel in the models above.
For release 4.0 we should focus on redesigning the UI.

**Reply by mokun** [reply]
> Science studies become a tab in the Monitor Tool and they have a dedicated Window when selected.

So, does it mean that the Science tab on Monitor Tool will be a global panel showing all the studies ?

Or it should be showing only the studies from one sponsor/settlement ? 

Now that we think about that, the Mission tab has been a global panel showing all the missions.

Are we going to reduce the cognitive load by allowing it to show only those missions per sponsor/settlement ?

We could add a checkbox or a radio button to switch between the sponsor view and the global view, can we ?

**Reply by bevans2000** [reply]
My suggestion is we do drop the global tools if they suit the monitor tab window. So we should scientific studies filtered on Settlement of the primary researcher.
For the Mission tool, we should drop it and use the existing monitor tab but have a new details window that is the right hand of the Mission tool. Same as scientific tool

**Reply by bevans2000** [reply]
I found this interesting Java library that we could use as a basis for a more dynamic UI docking solution https://github.com/andrewauclair/ModernDocking

**Reply by mokun** [reply]
> dynamic UI docking solution https://github.com/andrewauclair/ModernDocking

Looks like a good solution for us. We don't have to develop a docking mechanism from scratch.

**Reply by mokun** [reply]
>  we should scientific studies filtered on Settlement of the primary researcher.

We may as well first convert the combobox to sponsors before proceeding with this new change, right ?

**Reply by bevans2000** [reply]
Yes I agree, we should do the Authority change first in 3.8.0

**Reply by mokun** [reply]
btw, regarding authority (settlement/sponsor) combo box, I encounter two anomaly.

(1). If we are to start the sim as a standard scenario, 'Arcadia' will normally be the first top settlement to be listed in the authority (settlement/sponsor) combo box. 

However, let's say I open the settlement map on a settlement, say, 'Foundation'. All of a sudden, it would jump back to 'Arcadia' while I was still in the middle of viewing what's on that map. Why would that happen ?

(2). the authority combo normally start out alphabetically for all the settlement. However, after using it a while, it would suddenly become non-alphabetical, which makes it harder for me to find my settlement. 

Have you seen that ?

**Comment by bevans2000** [comment]
To move to a web UI the first step is to convert the current engine into a API service. Based on the graph nature GraphQL is a great choice. I propose we look to add an API as part of 3.9 with a roadmap of a simple web UI in 4.0. This is a long road so we need to support the Java UI inparallel

**Comment by bevans2000** [comment]
The issue of the UI came up in #1581. My thoughts are

1. JavaFX looks good but I don't know it and don't want to spend time learning it
2. WebUI is the perfect answer but there are so many things to do with this. I love ReactJS but don't fe el confident building a whole app like marssim on it
3. Swing desktop we have now and it works but can always be improved
4. Swing docking windows, ie IDE style, would be better and should allow us to reuse most of the panels we already have

Option 4 is the one that gets my vote. Option 2 is the correct long term but I don't have the time currently. So I think for 3.10 we consolidate the various back end improvements and set up 4.0 as a total rework of the front end.
Thoughts?

**Reply by Urwumpe** [reply]
I agree, 4 is the best option. Also I think that 2 has some merits, but that its better a project of its own, since I don't feel like a stable baseline could be achieved there anytime soon. A WebUI would likely see many changes and experiments until a state is achieved that can at least partially compete with Swing. Same for a Java based WebUI framework like Vaadin.

While I like JavaFX, I don't see it being automatically better than 2 or 4. And its a lot of effort to migrate, too.

3 feels a bit antique to use, but it does its job well and the effort from 3 to 4 shouldn't be so large that it can't be done in year.

What I could imagine as intermediate step into any direction, would be a internal framework or DSL to define new UI tabs or extension of tabs by defining the intended content, without breaking the look and feel or produce an unusual layout. This way, you could also make changes from 3 to 4 easier. Also, adding new features or optional extensions could be easier and less likely to break stuff.

**Reply by bevans2000** [reply]
There is a framework I found earlier in this conversation. It should be possible to take the base window we have for the desktop version and recode it to support the docketing framework. Hence many of the windows will get converted with little effort.
I feel a long term branch coming on with some experimentation.
But if this is accepted, then it would merit a change of majority version number to 4.0.

**Reply by Urwumpe** [reply]
IMHO, it would be nice if the UI can be better configured for what the user wants to do. Maybe with the possibility to define, export and import presets. AFAIR, most docking frameworks include support for this already.

And my opinion on versions: Lets talk about versions, when its done and integrated. Developing for a certain version feels so much like waterfall project management.

**Reply by bevans2000** [reply]
Had a deeper look into this. I think a parallel running approach will be best because it could be a long job with some experimentation. I see a slight reworking of the existing Swing UI Tool windows design. We should separate the panel content from the underlying window. Then the panel content can be reused in the new docking approach whilst still supporting the current UI.

**Reply by mokun** [reply]
I'd like to avoid disruption.

We're working on upgrading existing swing ui to become dockable.

Is there a way to keep the existing "Free Window" mode as an option in the Main Menu ?

Can we create an UI mode menu for players to choose what UI they prefer ?

**Reply by bevans2000** [reply]
Yes I think so. It would have to be a different main class or a command line argument

**Comment by mokun** [comment]
I just watched this short 10-min [video](https://www.youtube.com/watch?v=q39U9hp6gRw) on the future of smartphone UI.

It's about Google's Material 3 Expressive versus Apple's Liquid Glass.

They do seem to encapsulate two different schools of philosophy.

I feel like while our goal isn't about making our UI appealing, we should evaluate what usability goals we would like to see down the road and craft how we may achieve these goals.

**Comment by bevans2000** [comment]
@mokun @Urwumpe I've been thinking about this for the long term future. We should retain Java Swing because Web is too hard and JavaFX is a big re-write for not enough benefit.

The current Swing approach is a little inconsistent:

- The Mission, Science & Supply tools contain lists of Entities. But Monitor Tool is our place to show lists. Its confusing.
- Every Entity should have a dedicated window that just show the details. This window can be reached from multiple routes.
- The Internal frame approach is not a good use of screen estate. The content of the tools is tightly coupled to the windowing.

### Proposal

We should adopt the docking framework referenced above and allow users to position their screen where ever they want in a style similar to an IDE.
The content of the windows need to be separated from the windowing approach to give us flexibility. The docking window solution can then be added without a major change to the window content.

To remove the inconsistency, the Science & Resupply entities should have dedicated Entity windows and these 2 should be added to the Monitor window. The new Monitor window with the filtering fits this well.

To make the UI more accessible and easy to understand we should add an Entity Browser window than replaces the Search window. This allows the user to browse the Entities grouped by Settlements. Then all other Entities can be "owned" by a Settlement allowing the user to drill down the hierarchy in a tree like structure.

### Implementation

We can not just break everything so it has to be a gradual transition.
Key change to describe window content as a `ContentPanel` which will replace the current combined `ToolWindow`. A `ContentWindow` then provides a container a content panel to a physical window.
Furthermore a `UnitContentPanel` is specifically for showing the details of a `Unit` subclass; this will automatically attach as a UnitListener and forward events.

```mermaid
classDiagram
    JPanel <|-- ContentPanel
    class ContentPanel{
        <<abstract>>
        String name
        Icon icon
        +update()
        +destroy()
    }

    class TabPanel {
        String name
        Icon icon
    }
    TabPanel "many" *-- ContentPanel :  displays

    UIContext <.. ContentPanel
    class UIContext{
        <<interface>>

      +getSimulation() Simulation
      +getConfig() SimulationConfig
      +showDetails(Entity)
      +showLocation(Entity)
    }
    note for UIContext "Provides an interface to the underlying Simulation and Windowing"

    ContentPanel <|-- SettlementMapDisplay

    ContentPanel <|-- UnitContentPanel
    class UnitContentPanel{
            <<abstract>>
            Unit : monitored
            +unitUpdated()
            +destroy()
    }
	class UnitListener {
		<<interface>>
	}
	UnitListener <|-- UnitContentPanel
    UnitContentPanel <|-- PersonDetails
    note for PersonDetails "Rework of existing PersonWindow"
    UnitContentPanel <|-- SettlementDetails

    ContentPanel o-- ContentWindow
    UIContext <|-- MainDesktopWindow
    ModalInternalFrame <|-- ContentWindow
    ContentWindow "many" o-- MainDesktopWindow : displays

    namespace InternalWindowLayout {
        class ModalInternalFrame
        class ContentWindow
        class MainDesktopWindow
    }
```

**Reply by mokun** [reply]
> We can not just break everything so it has to be a gradual transition

JavaFX can run most, if not It all, Swing code.

**Reply by bevans2000** [reply]
If we adopt the pattern of `ContentPanel` we could render them in JavaFX I think. But I don't want to do it.
Ideally we should go for web,whilst I can do a GraphQL API, I'm not skilled enough to a JavaScript frontend using ReactJS I wish I could

**Reply by mokun** [reply]
> not skilled enough to a JavaScript frontend using ReactJS 

In terms of coming up with a decent UI design, we can defer our effort to an AI coding agents such as [Stitch](https://stitch.withgoogle.com/). See [this](https://www.youtube.com/watch?v=Gnnw4BYTpjY) for a quick intro.

**Comment by bevans2000** [comment]
I have created a Project to start shaping the changes to redo the Swing UI. I see this as a long term objective
https://github.com/orgs/mars-sim/projects/18

**Comment by bevans2000** [comment]
@mokun @Urwumpe  OK so I've implemented the framework described above making the visual content details independent of the underlying windowing framework. There is a PR #1848 
I like this approach and it's working well. There is some problem with double scrollbars sometimes that I've yet to hunt down. The Building, Settlement & Construction Unit Windows haven't been converted yet but they will be done soon. The existing UI still works fine with the new framework.
<img width="1919" height="1021" alt="image" src="https://github.com/user-attachments/assets/bc7fddfe-512a-400c-a8a4-eafe78f49e9f" />

**Comment by bevans2000** [comment]
Also I want to remove the JavaFX modules from teh master branch and just keep them on the JavaFX branch. The code is not used and it contributing to the code scanning and loaded into an IDE. It also can be confusing to a new contributors.

**Comment by bevans2000** [comment]
@mokun Ok it's done. I have completed the changes to create a docking style interface, I e. like an IDE

**Reply by mokun** [reply]
Great work !

**Reply by bevans2000** [reply]
There will be some snags. I am just working on generating more (& better) historical events to get more volume into the new Event Viewer. Should be done by weekend
